codec-http/src/main/java/io/netty/handler/codec/http/BackpressureHttpContentDecoder.java

```diff
@@ -0,0 +1,564 @@
+            if (HttpHeaderValues.IDENTITY.contentEquals(targetContentEncoding)) {
+                // Do NOT set the 'Content-Encoding' header if the target encoding is 'identity'
+                // as per: https://tools.ietf.org/html/rfc2616#section-14.11
+                headers.remove(HttpHeaderNames.CONTENT_ENCODING);
+            } else {
+                headers.set(HttpHeaderNames.CONTENT_ENCODING, targetContentEncoding);
+            }
+
+            readCompleteForHeaders = true;
+            if (message instanceof HttpContent) {
+                // If message is a full request or response object (headers + data), don't copy data part into out.
+                // Output headers only; data part will be decoded below.
+                // Note: "copy" object must not be an instance of LastHttpContent class,
+                // as this would (erroneously) indicate the end of the HttpMessage to other handlers.
+                HttpMessage copy;
+                if (message instanceof HttpRequest) {
+                    HttpRequest r = (HttpRequest) message; // HttpRequest or FullHttpRequest
+                    copy = new DefaultHttpRequest(r.protocolVersion(), r.method(), r.uri());
+                } else if (message instanceof HttpResponse) {
+                    HttpResponse r = (HttpResponse) message; // HttpResponse or FullHttpResponse
+                    copy = new DefaultHttpResponse(r.protocolVersion(), r.status());
+                } else {
+                    throw new CodecException("Object of class " + message.getClass().getName() +
+                            " is not an HttpRequest or HttpResponse");
+                }
+                copy.headers().set(message.headers());
+                copy.setDecoderResult(message.decoderResult());
+                needRead = false;
+                ctx.fireChannelRead(copy);
+            } else {
+                needRead = false;
+                ctx.fireChannelRead(message);
+                decoderHasReceivedInput = false;
+                return;
+            }
+        }
+
+        if (msg instanceof HttpContent) {
+            HttpContent c = (HttpContent) msg;
+            if (decoder == null) {
+                ctx.fireChannelRead(c);
+            } else {
+                decoderHasReceivedInput = true;
+                decoder.channelRead(decoderContext, c.content());
+
+                if (c instanceof LastHttpContent) {
+                    decoder.channelInactive(decoderContext);
+                    cleanup();
+
+                    LastHttpContent last = (LastHttpContent) c;
+                    // Generate an additional chunk if the decoder produced
+                    // the last product on closure,
+                    HttpHeaders headers = last.trailingHeaders();
+                    needRead = false;
+                    if (headers.isEmpty()) {
+                        ctx.fireChannelRead(LastHttpContent.EMPTY_LAST_CONTENT);
```

Comment:
just realized this is wrong, it needs to be delayed further. i will fix it tomorrow.